tests/test_connection.py

```diff
@@ -37,6 +39,51 @@ def test_connect_bad(conn_cls):
         conn_cls.connect("dbname=nosuchdb")
 
 
+def test_connect_error_single_host_original_message_preserved(conn_cls, monkeypatch):
+    # This is unreasonably short as it is expected to fail as fast as possible.
+    monkeypatch.setattr(psycopg.conninfo, "_DEFAULT_CONNECT_TIMEOUT", 0.1)
+
+    # IPv4 address blocks reserved for documentation.
+    # https://datatracker.ietf.org/doc/rfc5737/
+    args = {"host": "192.0.2.1", "port": "1234"}
+    with pytest.raises(psycopg.OperationalError) as e:
+        conn_cls.connect(**args)
+
+    msg = str(e)
+    assert "connection timeout expired" in msg
+    assert MULTI_FAILURE_MESSAGE not in msg
+
+
+def test_connect_error_multi_hosts_each_message_preserved(conn_cls, monkeypatch):
+    # This is unreasonably short as it is expected to fail as fast as possible.
+    monkeypatch.setattr(psycopg.conninfo, "_DEFAULT_CONNECT_TIMEOUT", 0.1)
+
+    # IPv4 address blocks reserved for documentation.
+    # https://datatracker.ietf.org/doc/rfc5737/
+    args = {"host": "192.0.2.1,198.51.100.1", "port": "1234,5678"}
```

Comment:
This is cool, I didn't know about these unroutable addresses! Thank you for the reference.

So far, in order to test with a host that doesn't reply, we opened a listening socket with [proxy.deaf_port](https://github.com/psycopg/psycopg/blob/24ad0eb967188702b6fccabb2d1b46a7f6a8b7c7/tests/fix_proxy.py#L105-L118) but if these IPs work reliably we can probably do without.

You can set `connect_timeout` in the connection string instead of monkeypatching. However, 0.1s is a slow test :slightly_smiling_face: so we need to mark it so.

```suggestion
@pytest.mark.slow
def test_connect_error_multi_hosts_each_message_preserved(conn_cls):
    # IPv4 address blocks reserved for documentation.
    # https://datatracker.ietf.org/doc/rfc5737/
    args = {"host": "192.0.2.1,198.51.100.1", "port": "1234,5678", "connect_timeout": 0.1}
```

same for both tests, both sync and async.

Reply:
I didn't use the proxy in this case since I needed 2 hosts and 2 ports that would reliably fail and it seemed like the proxy only exposed 1 of each. If you have a suggestion for how to do that with the proxy then I'm happy to swap, but the documentation IPs should work reliably and have done so in my testing.

I'll remove the monkeypatch and mark the tests accordingly, thanks for the tips on this.

Reply:
These tests hosts are ok. In other tests we have in the test suite probably we monkeypatch the dns to inject expected failures, but I don't think it's needed here.

Reply:
Should be addressed by [24c2995](https://github.com/psycopg/psycopg/commit/24c299561e3f13793029fa98706aaac637fce92d)